adr/ADR-42.md

```diff
@@ -0,0 +1,168 @@
+providing exact control would not be in the interest of distributed patterns, so instead we'll support a concept of priorities
+between 1 and 10. Multiple clients can share the same priority which will result in a random client be picked from the lowest
+priority as the next candidate for pinning. If no priority is supplied in a pull it will default to 5.
+
+**NOTE: We should not describe this in terms of exclusivity as there is no such guarantee, there will be times when one
+client think it is pinned when it is not because the server switched.** 
+
+If the current pinned client is priority 5 and a pull comes in for this group from a lower priority client the server will
+switch to this lower priority client.
+
+The `pinned_client` policy provides server-side orchestration for the selection of the pinned client.
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "pinned_client",
+	PriorityTimeout: 120*time.Second,
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+This configuration states:
+
+ * We have 1 group defined and all pulls have to belong to this group
+ * The policy is `pinned_client` that activates these behaviors
+ * When a pinned client has not done any pulls in the last 120 seconds the server will switch to another client
+ * AckPolicy has to be `explicit`. If we force this ack policy in normal use we should error in Pedantic mode
+
+A pull request will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"priority": 1` - the priority this pull is done at, valid values are 1 to 10. Absent means 5. Anything else is an error
+ * `"id": "xyz"` - the pinned client will have this ID set to the one the server supplied (see below), otherwise this field is absent
+
+When no pinned client has been selected by the server the first message that will be delivered will include a `Nats-Pin: xyz`
+header. The client that gets this message should at that point ensure that all future pull requests have the same ID set.
+The pinned client is chosen by walking all the waiting pulls and picking the lowest priority one.
+
+If a lower priority pull is received than the current pinned clients most recent pull the server will switch to a new
+pinned client:
+
+ 1. Stop delivering messages for this group, wait for all in-flight messages to be completed, continue to serve heartbeats etc 
+ 2. Pick the new pinned client by priority and group
+ 3. Store the new pinned `nuid`
+ 4. Deliver the message to the new pinned client with the ID set
+ 5. Create an advisory that a new pinned client was picked
+ 6. Respond with a 4xx header to any pulls, including waiting ones, that have a different ID set. Client that received this error will clear the ID and pull with no ID 
+
+If no pulls from the pinned client is received within `PriorityTimeout` the server will switch again using the same flow as above.
```

Comment:
I'm pretty sure rather quickly someone will ask for graceful, scheduled handover of the pinned consumer to not cause any (or almost any) interrupts in opeartions.

Maybe this could be addressed with `consume()` like pattern and very low PriorityTimeout. Or even 0, as `consume()` always has at least one pending Fetch Request.

Reply:
Yeah, I think we just don't know enough about the patterns users will use this for - if we build a whole solution ontop of this ourselves like the one we are planning we could definitely do that - but I reckon for sute this tunable have to exist I can't imagine a good number for everyone since I suspect people will use these for all sorts of stuff that we can't even anticipate.